Vite: Move mdx-plugin from @storybook/builder-vite to @storybook/addon-docs

bryanjtc opened the pull request:
Closes #24058

<!-- If your PR is related to an issue, provide the number(s) above; if it resolves multiple issues, be sure to break them up (e.g. "closes #1000, closes #1001"). -->

<!--

Thank you for contributing to Storybook! Please submit all PRs to the `next` branch unless they are specific to the current release. Storybook maintainers cherry-pick bug and documentation fixes into the `main` branch as part of the release process, so you shouldn't need to worry about this. For additional guidance: https://storybook.js.org/docs/react/contribute/how-to-contribute

-->


## What I did

Moved the mdx plugin from the vite builder to addon docs.

## Checklist for Contributors

### Testing

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to communicate how to test your changes -->

#### The changes in this PR are covered in the following automated tests:
- [ ] stories
- [ ] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

No manual test needed

<!-- Please include the steps to test your changes here. For example:

1. Run a sandbox for template, e.g. `yarn task --task sandbox --start-from auto --template react-vite/default-ts`
2. Open Storybook in your browser
3. Access X story

-->

### Documentation

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to indicate which documentation has been updated. -->

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [x] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli/src/sandbox-templates.ts`
- [x] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release

<!-- CANARY_RELEASE_SECTION -->

This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=<PR_NUMBER>`_

<!-- CANARY_RELEASE_SECTION -->

bryanjtc committed:
fix(builder-vite): Fix missing @storybook/addon-docs dependency

shilman reviewed (commented):
can we solve this a different way by moving the MDX code out of the vite builder and into addon-docs viteFinal method? @ndelangen

IanVS commented:
That feels like the right way to do it to me, @shilman.

ndelangen commented:
I agree, @bryanjtc is this something you could do?

bryanjtc commented:
sure

bryanjtc committed:
Move mdx plugin from builder-vite to addon-docs

bryanjtc committed:
Await mdxPlugin

bryanjtc committed:
Replace createFilter import from vite with @rollup/pluginutils

bryanjtc committed:
Use StorybookConfig type from @storybook/builder-vite

bryanjtc committed:
Replace parameter for mdxPlugin

bryanjtc committed:
Move mdxPlugin import

bryanjtc committed:
Replace types in viteFinal to fix error

bryanjtc commented:
The recent release of @testing-library/user-event version 4.5.0 has caused issues with @storybook/testing-library, resulting in failures in the sandboxes. I did a local test by downgrading to version 14.4.3 in the react-vite-default-js sandbox, and I can confirm that the code now runs without encountering any errors.

IanVS commented on code/addons/docs/src/plugins/mdx-plugin.ts:
```diff
@@ -34,10 +33,7 @@ export async function mdxPlugin(options: Options): Promise<Plugin> {
-            dirname(require.resolve('@storybook/addon-docs/package.json')),
-            '/dist/shims/mdx-react-shim'
-          ),
+          providerImportSource: require.resolve('@storybook/mdx2-csf'),
```
The reason for the funky join-with-dirname is to allow this to create an extensionless import.  If you simply `require.resolve` like this, the result will use the `require` condition from `@storybook/mdx2-csf`, and will point to the cjs entrypoint, even in the esm build.  So I think we need to keep using the more complicated strategy here.

bryanjtc replied:
Since the plugin is now inside `@storybook/addons-docs`, should I just use the direct path to the package.json? 
Like this:
```ts
dirname(require.resolve('../../package.json')),'/dist/shims/mdx-react-shim'
```

IanVS replied:
Yeah I think so, what do you say, @ndelangen?

IanVS reviewed (commented):
Thanks so much for tackling this!  I've wanted to move that plugin out of the vite builder for a long time, just never got around to it.

bryanjtc committed:
Keep the previous providerImportSource value

ndelangen commented on code/addons/docs/package.json:
```diff
@@ -100,6 +100,7 @@
   "dependencies": {
     "@jest/transform": "^29.3.1",
     "@mdx-js/react": "^2.1.5",
+    "@rollup/pluginutils": "^5.0.2",
```
Could this be moved to a devDependency (and thus be bundled in)?, therefore not expose this dependency to users.

ndelangen replied:
Or is this a peerDependency in some way?

bryanjtc committed:
Move @rollup/pluginutils to devDependencies

bryanjtc commented:
> @bryanjtc are you confident that this truly needs no manual regression test?

The mdx plugin is used only to load mdx in vite projects. Running all vite sandboxes successfully should be enough.

ndelangen commented:
I'll defer to @shilman and @JReinhold to proceed to merge this.

bryanjtc commented:
> This looks like a good change to me. The only additional thing I can think of that might be worth testing is a vite project that does not have `addon-docs` installed, since I don't think any of the sandboxes are configured that way. Presumably that wouldn't cause any problems, because then mdx wouldn't be used in the project and this vite plugin wouldn't be needed.

I did a manual test modifying the package.json and main.ts from the react-vite-default-ts sandbox and it works.
I removed addon-docs and addon-essentials from the package.json.
Then removed mdx stories from from the main.ts and added the other addons that come with addon-essentials.

```main.ts
import type { StorybookConfig } from '@storybook/react-vite';

const config: StorybookConfig = {
  stories: ['../src/**/*.stories.@(js|jsx|mjs|ts|tsx)', {
    directory: '../template-stories/lib/preview-api',
    titlePrefix: 'lib/preview-api',
    files: '**/*.@(stories.@(js|jsx|ts|tsx))'
  }, {
    directory: '../template-stories/addons/links',
    titlePrefix: 'addons/links',
    files: '**/*.@(stories.@(js|jsx|ts|tsx))'
  }, {
    directory: '../template-stories/addons/actions',
    titlePrefix: 'addons/actions',
    files: '**/*.@(stories.@(js|jsx|ts|tsx))'
  }, {
    directory: '../template-stories/addons/backgrounds',
    titlePrefix: 'addons/backgrounds',
    files: '**/*.@(stories.@(js|jsx|ts|tsx))'
  }, {
    directory: '../template-stories/addons/controls',
    titlePrefix: 'addons/controls',
    files: '**/*.@(stories.@(js|jsx|ts|tsx))'
  }, {
    directory: '../template-stories/addons/docs',
    titlePrefix: 'addons/docs',
    files: '**/*.@(stories.@(js|jsx|ts|tsx))'
  }, {
    directory: '../template-stories/addons/toolbars',
    titlePrefix: 'addons/toolbars',
    files: '**/*.@(stories.@(js|jsx|ts|tsx))'
  }, {
    directory: '../template-stories/addons/viewport',
    titlePrefix: 'addons/viewport',
    files: '**/*.@(stories.@(js|jsx|ts|tsx))'
  }, {
    directory: '../template-stories/addons/interactions',
    titlePrefix: 'addons/interactions',
    files: '**/*.@(stories.@(js|jsx|ts|tsx))'
  }],

  addons: [
    '@storybook/addon-links',
    '@storybook/addon-onboarding',
    '@storybook/addon-interactions',
    '@storybook/addon-actions',
    '@storybook/addon-backgrounds',
    '@storybook/addon-controls',
    '@storybook/addon-highlight',
    '@storybook/addon-outline',
    '@storybook/addon-toolbars',
    '@storybook/addon-viewport',
  ],

  framework: {
    name: '@storybook/react-vite',
    options: {},
  },

  docs: {
    autodocs: 'tag',
  },

  previewAnnotations: ['./src/stories/components', './template-stories/lib/preview-api/preview.ts', './template-stories/addons/toolbars/preview.ts'],
  features: {},

  core: {
    disableWhatsNewNotifications: true
  },

  viteFinal: (config) => ({
    ...config,
    optimizeDeps: { ...config.optimizeDeps, force: true },
    server: {
      ...config.server,
      fs: {
        ...config.server?.fs,
        allow: ['stories', 'src', 'template-stories', 'node_modules', ...(config.server?.fs?.allow || [])],
      },
    },
  })
};
export default config;
```

package.json
```json
{
  "name": "before-storybook",
  "private": true,
  "version": "0.0.0",
  "type": "module",
  "scripts": {
    "dev": "vite",
    "build": "tsc && vite build",
    "lint": "eslint . --ext ts,tsx --report-unused-disable-directives --max-warnings 0",
    "preview": "vite preview",
    "storybook": "NODE_OPTIONS='--preserve-symlinks --preserve-symlinks-main' STORYBOOK_TELEMETRY_URL=\"http://localhost:6007/event-log\" storybook dev -p 6006",
    "build-storybook": "NODE_OPTIONS='--preserve-symlinks --preserve-symlinks-main' STORYBOOK_TELEMETRY_URL=\"http://localhost:6007/event-log\" storybook build"
  },
  "dependencies": {
    "@types/node": "16",
    "react": "^18.2.0",
    "react-dom": "^18.2.0"
  },
  "devDependencies": {
    "@storybook/addon-actions": "^7.5.0-alpha.2",
    "@storybook/addon-backgrounds": "^7.5.0-alpha.2",
    "@storybook/addon-controls": "^7.5.0-alpha.2",
    "@storybook/addon-highlight": "^7.5.0-alpha.2",
    "@storybook/addon-interactions": "^7.5.0-alpha.2",
    "@storybook/addon-links": "^7.5.0-alpha.2",
    "@storybook/addon-measure": "^7.5.0-alpha.2",
    "@storybook/addon-onboarding": "^1.0.8",
    "@storybook/addon-outline": "^7.5.0-alpha.2",
    "@storybook/addon-toolbars": "^7.5.0-alpha.2",
    "@storybook/addon-viewport": "^7.5.0-alpha.2",
    "@storybook/blocks": "^7.5.0-alpha.2",
    "@storybook/jest": "next",
    "@storybook/react": "^7.5.0-alpha.2",
    "@storybook/react-vite": "^7.5.0-alpha.2",
    "@storybook/test-runner": "next",
    "@storybook/testing-library": "next",
    "@types/react": "^18.2.15",
    "@types/react-dom": "^18.2.7",
    "@typescript-eslint/eslint-plugin": "^6.0.0",
    "@typescript-eslint/parser": "^6.0.0",
    "@vitejs/plugin-react": "^4.0.3",
    "eslint": "^8.45.0",
    "eslint-plugin-react-hooks": "^4.6.0",
    "eslint-plugin-react-refresh": "^0.4.3",
    "eslint-plugin-storybook": "^0.6.13",
    "storybook": "^7.5.0-alpha.2",
    "typescript": "^5.0.2",
    "vite": "^4.4.5"
  },
  "packageManager": "yarn@3.6.3",
  "resolutions": {
    "@storybook/addon-a11y": "portal:/home/bryan/projects/storybook/code/addons/a11y",
    "@storybook/addon-actions": "portal:/home/bryan/projects/storybook/code/addons/actions",
    "@storybook/addon-backgrounds": "portal:/home/bryan/projects/storybook/code/addons/backgrounds",
    "@storybook/addon-controls": "portal:/home/bryan/projects/storybook/code/addons/controls",
    "@storybook/addon-mdx-gfm": "portal:/home/bryan/projects/storybook/code/addons/gfm",
    "@storybook/addon-highlight": "portal:/home/bryan/projects/storybook/code/addons/highlight",
    "@storybook/addon-interactions": "portal:/home/bryan/projects/storybook/code/addons/interactions",
    "@storybook/addon-jest": "portal:/home/bryan/projects/storybook/code/addons/jest",
    "@storybook/addon-links": "portal:/home/bryan/projects/storybook/code/addons/links",
    "@storybook/addon-measure": "portal:/home/bryan/projects/storybook/code/addons/measure",
    "@storybook/addon-outline": "portal:/home/bryan/projects/storybook/code/addons/outline",
    "@storybook/addon-storyshots": "portal:/home/bryan/projects/storybook/code/addons/storyshots-core",
    "@storybook/addon-storyshots-puppeteer": "portal:/home/bryan/projects/storybook/code/addons/storyshots-puppeteer",
    "@storybook/addon-storysource": "portal:/home/bryan/projects/storybook/code/addons/storysource",
    "@storybook/addon-themes": "portal:/home/bryan/projects/storybook/code/addons/themes",
    "@storybook/addon-toolbars": "portal:/home/bryan/projects/storybook/code/addons/toolbars",
    "@storybook/addon-viewport": "portal:/home/bryan/projects/storybook/code/addons/viewport",
    "@storybook/builder-manager": "portal:/home/bryan/projects/storybook/code/builders/builder-manager",
    "@storybook/builder-vite": "portal:/home/bryan/projects/storybook/code/builders/builder-vite",
    "@storybook/builder-webpack5": "portal:/home/bryan/projects/storybook/code/builders/builder-webpack5",
    "@storybook/addons": "portal:/home/bryan/projects/storybook/code/deprecated/addons",
    "@storybook/channel-postmessage": "portal:/home/bryan/projects/storybook/code/deprecated/channel-postmessage",
    "@storybook/channel-websocket": "portal:/home/bryan/projects/storybook/code/deprecated/channel-websocket",
    "@storybook/client-api": "portal:/home/bryan/projects/storybook/code/deprecated/client-api",
    "@storybook/core-client": "portal:/home/bryan/projects/storybook/code/deprecated/core-client",
    "@storybook/api": "portal:/home/bryan/projects/storybook/code/deprecated/manager-api-shim",
    "@storybook/preview-web": "portal:/home/bryan/projects/storybook/code/deprecated/preview-web",
    "@storybook/store": "portal:/home/bryan/projects/storybook/code/deprecated/store",
    "@storybook/angular": "portal:/home/bryan/projects/storybook/code/frameworks/angular",
    "@storybook/ember": "portal:/home/bryan/projects/storybook/code/frameworks/ember",
    "@storybook/html-vite": "portal:/home/bryan/projects/storybook/code/frameworks/html-vite",
    "@storybook/html-webpack5": "portal:/home/bryan/projects/storybook/code/frameworks/html-webpack5",
    "@storybook/nextjs": "portal:/home/bryan/projects/storybook/code/frameworks/nextjs",
    "@storybook/preact-vite": "portal:/home/bryan/projects/storybook/code/frameworks/preact-vite",
    "@storybook/preact-webpack5": "portal:/home/bryan/projects/storybook/code/frameworks/preact-webpack5",
    "@storybook/react-vite": "portal:/home/bryan/projects/storybook/code/frameworks/react-vite",
    "@storybook/react-webpack5": "portal:/home/bryan/projects/storybook/code/frameworks/react-webpack5",
    "@storybook/server-webpack5": "portal:/home/bryan/projects/storybook/code/frameworks/server-webpack5",
    "@storybook/svelte-vite": "portal:/home/bryan/projects/storybook/code/frameworks/svelte-vite",
    "@storybook/svelte-webpack5": "portal:/home/bryan/projects/storybook/code/frameworks/svelte-webpack5",
    "@storybook/sveltekit": "portal:/home/bryan/projects/storybook/code/frameworks/sveltekit",
    "@storybook/vue-vite": "portal:/home/bryan/projects/storybook/code/frameworks/vue-vite",
    "@storybook/vue-webpack5": "portal:/home/bryan/projects/storybook/code/frameworks/vue-webpack5",
    "@storybook/vue3-vite": "portal:/home/bryan/projects/storybook/code/frameworks/vue3-vite",
    "@storybook/vue3-webpack5": "portal:/home/bryan/projects/storybook/code/frameworks/vue3-webpack5",
    "@storybook/web-components-vite": "portal:/home/bryan/projects/storybook/code/frameworks/web-components-vite",
    "@storybook/web-components-webpack5": "portal:/home/bryan/projects/storybook/code/frameworks/web-components-webpack5",
    "@storybook/channels": "portal:/home/bryan/projects/storybook/code/lib/channels",
    "@storybook/cli": "portal:/home/bryan/projects/storybook/code/lib/cli",
    "sb": "portal:/home/bryan/projects/storybook/code/lib/cli-sb",
    "storybook": "portal:/home/bryan/projects/storybook/code/lib/cli-storybook",
    "@storybook/client-logger": "portal:/home/bryan/projects/storybook/code/lib/client-logger",
    "@storybook/codemod": "portal:/home/bryan/projects/storybook/code/lib/codemod",
    "@storybook/core-common": "portal:/home/bryan/projects/storybook/code/lib/core-common",
    "@storybook/core-events": "portal:/home/bryan/projects/storybook/code/lib/core-events",
    "@storybook/core-server": "portal:/home/bryan/projects/storybook/code/lib/core-server",
    "@storybook/core-webpack": "portal:/home/bryan/projects/storybook/code/lib/core-webpack",
    "@storybook/csf-plugin": "portal:/home/bryan/projects/storybook/code/lib/csf-plugin",
    "@storybook/csf-tools": "portal:/home/bryan/projects/storybook/code/lib/csf-tools",
    "@storybook/docs-tools": "portal:/home/bryan/projects/storybook/code/lib/docs-tools",
    "@storybook/instrumenter": "portal:/home/bryan/projects/storybook/code/lib/instrumenter",
    "@storybook/manager-api": "portal:/home/bryan/projects/storybook/code/lib/manager-api",
    "@storybook/node-logger": "portal:/home/bryan/projects/storybook/code/lib/node-logger",
    "@storybook/postinstall": "portal:/home/bryan/projects/storybook/code/lib/postinstall",
    "@storybook/preview": "portal:/home/bryan/projects/storybook/code/lib/preview",
    "@storybook/preview-api": "portal:/home/bryan/projects/storybook/code/lib/preview-api",
    "@storybook/react-dom-shim": "portal:/home/bryan/projects/storybook/code/lib/react-dom-shim",
    "@storybook/router": "portal:/home/bryan/projects/storybook/code/lib/router",
    "@storybook/source-loader": "portal:/home/bryan/projects/storybook/code/lib/source-loader",
    "@storybook/telemetry": "portal:/home/bryan/projects/storybook/code/lib/telemetry",
    "@storybook/theming": "portal:/home/bryan/projects/storybook/code/lib/theming",
    "@storybook/types": "portal:/home/bryan/projects/storybook/code/lib/types",
    "@storybook/preset-create-react-app": "portal:/home/bryan/projects/storybook/code/presets/create-react-app",
    "@storybook/preset-html-webpack": "portal:/home/bryan/projects/storybook/code/presets/html-webpack",
    "@storybook/preset-preact-webpack": "portal:/home/bryan/projects/storybook/code/presets/preact-webpack",
    "@storybook/preset-react-webpack": "portal:/home/bryan/projects/storybook/code/presets/react-webpack",
    "@storybook/preset-server-webpack": "portal:/home/bryan/projects/storybook/code/presets/server-webpack",
    "@storybook/preset-svelte-webpack": "portal:/home/bryan/projects/storybook/code/presets/svelte-webpack",
    "@storybook/preset-vue-webpack": "portal:/home/bryan/projects/storybook/code/presets/vue-webpack",
    "@storybook/preset-vue3-webpack": "portal:/home/bryan/projects/storybook/code/presets/vue3-webpack",
    "@storybook/preset-web-components-webpack": "portal:/home/bryan/projects/storybook/code/presets/web-components-webpack",
    "@storybook/html": "portal:/home/bryan/projects/storybook/code/renderers/html",
    "@storybook/preact": "portal:/home/bryan/projects/storybook/code/renderers/preact",
    "@storybook/react": "portal:/home/bryan/projects/storybook/code/renderers/react",
    "@storybook/server": "portal:/home/bryan/projects/storybook/code/renderers/server",
    "@storybook/svelte": "portal:/home/bryan/projects/storybook/code/renderers/svelte",
    "@storybook/vue": "portal:/home/bryan/projects/storybook/code/renderers/vue",
    "@storybook/vue3": "portal:/home/bryan/projects/storybook/code/renderers/vue3",
    "@storybook/web-components": "portal:/home/bryan/projects/storybook/code/renderers/web-components",
    "@storybook/blocks": "portal:/home/bryan/projects/storybook/code/ui/blocks",
    "@storybook/components": "portal:/home/bryan/projects/storybook/code/ui/components",
    "@storybook/manager": "portal:/home/bryan/projects/storybook/code/ui/manager",
    "@vitejs/plugin-react": "^4.0.0"
  }
}

```

bryanjtc commented:
Can you merge this with 7.5.0?

IanVS commented:
I'd merge it, but it sounds like Norbert wants to leave it up to @shilman or @JReinhold, so I've assigned them.

JReinhold commented on code/addons/docs/package.json:
```diff
@@ -119,9 +119,11 @@
   },
   "devDependencies": {
+    "@rollup/pluginutils": "^5.0.2",
     "react": "^16.14.0",
     "react-dom": "^16.8.0",
-    "typescript": "~4.9.3"
+    "typescript": "~4.9.3",
+    "vite": "^4.0.4"
```
@ndelangen can you confirm here that this means `@rollup/pluginutils` will get prebundled in, but `vite` and `typescript` won't, because they're not imported in the source?

It's roughly a [23kB increase in bundle size](https://bundlejs.com/?q=%40rollup%2Fpluginutils&treeshake=%5B%7B+createFilter+%7D%5D), but I don't see any other way, even source-copy is the same size

ndelangen replied:
Yes.

ndelangen commented:
MERGED 🎉
